pandas/core/frame.py

```diff
@@ -3730,7 +3745,13 @@ def reindexer(value):
             infer_dtype, _ = infer_dtype_from_scalar(value, pandas_dtype=True)
 
             # upcast
-            value = cast_scalar_to_array(len(self.index), value)
+            if is_extension_array_dtype(infer_dtype):
+                value = construct_1d_arraylike_from_scalar(
+                    value, len(self.index), infer_dtype
+                )
+            else:
+                value = cast_scalar_to_array(len(self.index), value)
```

Comment:
can you not do this *inside* cast_scalar_to_array (pass in the `infer_dtype` optionally)

Reply:
Can you clarify what you mean by "this"? Your last comments were to move this whole if/else logic out of `cast_scalar_to_array`, which I do agree made for a much cleaner solution

Reply:
instead of adding this if/then clause, can you change `cast_scalar_to_array` to just handle if the dtype is an extension_array? (and obviously pass infer_dtype here)

Reply:
`cast_scalar_to_array` only deals with numpy arrays. Initially this PR added handling extension dtypes in there as well, but after discussion we actually moved it out because it gave a lot of complications (eg EAs don't support 2D arrays, while `cast_scalar_to_array` generally does).

Reply:
exactly the reason to make the change. we are almost certainly going to need to do this in other places and this is a natural fit.

Reply:
@jreback then please first answer your own comment at https://github.com/pandas-dev/pandas/pull/34875#discussion_r445880416 on how to solve it differently (and how to deal with 2D shape in `cast_scalar_to_array` in case of extension dtypes). Moving it out of `cast_scalar_to_array` was exactly to avoid this "crazy api" you commented on.

Reply:
I agree that that would be ideal, but I believe that would be a much larger refactor and to me it seems like something that should have it's own PR.

I do acknowledge that this PR is a band-aid that may not be required if we were to make that change. But I still think it would be best to apply the band-aid and come back to that at a later time given the maturity of this PR and the complexity associated with that refactor.

Though if this is something that we want to put in this PR, I would benefit from some fairly verbose assistance in coding that up. I'm not sure that I have the context and code familiarity to do that alone.

Reply:
@jreback What are your thoughts here?

Reply:
how does this involve 2D at all here? simply move the if_extenseion array case *into* `cast_scalar_arry`. The constructor here is WAY to big to grok, adding if/else cases makes this much worse.

Reply:
``cast_scalar_array`` in general involves 2D (not here, but it is also used a bit higher in this file in the DataFrame constructor with a 2D shape, and a similar if/else structure is used there). If we let it handle extension arrays, we should let it handle those for all shapes IMO, otherwise it is making that only more confusing.

Again, please look at your own earlier comment (https://github.com/pandas-dev/pandas/pull/34875#discussion_r445880416 and the answers to that) and propose a way how to deal with this in `cast_scalar_array` without a "crazy api" (which was returning a list of 1D EAs for the 2D case).

Reply:
If you want to see the state at the point when things were handled inside `cast_scalar_array`, this is the diff of the commits before it was refactored: https://github.com/pandas-dev/pandas/pull/34875/files/90a85706ce1bfa1735cf07190dc61efaa62e454d

Reply:
I see the code and this is easily possible. please try to do as I have described. I don't see any reason not to.

Reply:
@jreback Please be more specific. What I would do is exactly as I had it, which you called a "crazy api" and specifically requested I change.
Are you asking me to revert the changes that you requested? Or are you after something else?

Reply:
To be more specific, it would be easy if `cast_scalar_array` was simply returning a single 1d array. But in most cases it is not, it is returning a 2d array.
Since there's no good way to make a 2d Extension array, I handled this by returning a list of Extension arrays. Is this what your asking I go back to doing?

Reply:
We cannot "simply move the if_extenseion array case into cast_scalar_arry", because this is a specific case where `cast_scalar_array` is receiving an int to its `shape` parameter. But in general, `shape` is a tuple to define the shape of the 2d array.

Reply:
https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1492

you just need to move the if_extension_dtype *into* this function. I don't think there is ambiguity or anything. This returns exactly 1 ndarray. I am not sure where 2D even matters here at all.

Reply:
A single ndarray can be 2 dimensional, and in this case it often is, for example, when it is used [here].(https://github.com/pandas-dev/pandas/blob/c73a2dee55f48690ad9a442ba019f9bf3bc22c26/pandas/core/frame.py#L552)

Doing this in `cast_scalar_to_array` would NOT work without a larger refactor.
```
if is_extension_array_dtype(dtype):
    value = construct_1d_arraylike_from_scalar(fill_value, shape, dtype)
else:
    values = np.empty(shape, dtype=dtype)
    values.fill(fill_value)
```
this `shape` parameter is usually a **tuple** not an **int** like it is here, so `construct_1d_arraylike_from_scalar` would fail when `shape=(2, 5)` for example (since extension arrays are only 1d)

Reply:
"This returns exactly 1 ndarray" is technically true, but it can be a 2d ndarray

For example
```python
values = np.empty((2,3), dtype=int)
values.fill(2)
```
would give you a numpy array like
```
[[2, 2, 2], [2, 2, 2]]
```